src/main/java/com/microsoft/sqlserver/jdbc/SQLServerSecurityUtility.java

```diff
@@ -319,191 +319,109 @@ static void verifyColumnMasterKeyMetadata(SQLServerConnection connection, SQLSer
-        String cacheKey = "resource:" + resource + "|clientid:" + msiClientId;
-        SqlFedAuthToken token = msiTokenCache.get(cacheKey);
-        if (token != null) {
-            if (connectionlogger.isLoggable(Level.FINER)) {
-                connectionlogger.finer("Using cached Managed Identity auth token: " + token.toString());
-            }
-            return token;
+    static SqlFedAuthToken getManagedIdentityCredAuthToken(String resource, String msiClientId) throws SQLServerException {
+        ManagedIdentityCredential mic = null;
+
+        if (null != msiClientId && !msiClientId.isEmpty()) {
+            mic = new ManagedIdentityCredentialBuilder().clientId(msiClientId).build();
+        } else {
+            mic = new ManagedIdentityCredentialBuilder().build();
         }
 
-        // IMDS upgrade time can take up to 70s
-        final int imdsUpgradeTimeInMs = 70 * 1000;
-        final List<Integer> retrySlots = new ArrayList<>();
+        TokenRequestContext tokenRequestContext = new TokenRequestContext();
+        String scope = resource.endsWith(SQLServerMSAL4JUtils.SLASH_DEFAULT) ? resource : resource + SQLServerMSAL4JUtils.SLASH_DEFAULT;
+        tokenRequestContext.setScopes(Arrays.asList(scope));
 
-        StringBuilder urlString = new StringBuilder();
-        int retry = 1, maxRetry = 1;
+        SqlFedAuthToken sqlFedAuthToken = null;
 
-        // MSI_ENDPOINT and MSI_SECRET can be used instead of IDENTITY_ENDPOINT and IDENTITY_HEADER
-        String identityEndpoint = System.getenv("IDENTITY_ENDPOINT");
-        if (null == identityEndpoint || identityEndpoint.trim().isEmpty()) {
-            identityEndpoint = System.getenv("MSI_ENDPOINT");
-        }
+        Optional<AccessToken> accessTokenOptional = mic
+                .getToken(tokenRequestContext)
+                .blockOptional();
 
-        String identityHeader = System.getenv("IDENTITY_HEADER");
-        if (null == identityHeader || identityHeader.trim().isEmpty()) {
-            identityHeader = System.getenv("MSI_SECRET");
-        }
+        if (!accessTokenOptional.isPresent()) {
+            if (connectionlogger.isLoggable(java.util.logging.Level.FINE)) {
+                connectionlogger.fine("Access token is not present.");
+            }
 
-        /*
-         * isAzureFunction is used for identifying if the current client application is running in a Virtual Machine
-         * (without Managed Identity environment variables) or App Service/Function (with Managed Identity environment
-         * variables) as the APIs to be called for acquiring MSI Token are different for both cases.
-         */
-        boolean isAzureFunction = null != identityEndpoint && !identityEndpoint.isEmpty() && null != identityHeader
-                && !identityHeader.isEmpty();
+            throw new SQLServerException(SQLServerException
+                    .getErrString("R_ManagedIdentityTokenAcquisitionFail"), null);

```

Comment:
can we add more info to the error message to make it easier to debug when there's a failure

Reply:
Similarly, do we have more info if the Optional task fails or doesn't return a result? I don't know the API well...

Reply:
Not sure. From the sounds of it, we don't have anything more available. The current iteration looks good.